sycl/doc/extensions/proposed/sycl_ext_oneapi_virtual_functions.asciidoc

```diff
@@ -0,0 +1,683 @@
+`aspect::ext_oneapi_virtual_functions`.  The application must check that the
+device has this aspect before submitting a kernel using any of the APIs in this
+extension.  If the application fails to do this, the implementation throws
+a synchronous exception with the `errc::kernel_not_supported` error code
+when the kernel is submitted to the queue.
+
+== Overview
+
+The main purpose of this extension is to reduce amount of SYCL language
+restrictions for device code by allowing to call virtual member functions
+from device functions.
+
+NOTE: this extension **does not** cover (i.e. doesn't enable) things like
+`dynamic_cast`, `typeid` or calls through function pointers.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_VIRTUAL_FUNCTIONS` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New language restrictions for device functions
+
+The following restriction, listed in section 5.4 of the core SYCL specification
+does not apply if this extension is supported by an implementation:
+
+> The odr-use of polymorphic classes and classes with virtual inheritance is
+> allowed. *However, no virtual member functions are allowed to be called in a
+> device function.*
+
+However, there are still some limitations of how virtual member functions can
+be used:
+
+- if an object is constructed in host code, calling a virtual member function
+  for that object in device code has undefined behavior;
+- if an object is constructed in device code on a device `A`, calling a virtual
+  member function for that object in host code, or on another device `B` has
+  undefined behavior;
+
+=== New properties
+
+Under the hood virtual functions are essentially function pointers which are
+stored in a global variable and managed by compiler-generated code. Therefore,
+each call to a virtual member function is an indirect call and compiler may not
+be able to understand which exact virtual function is being called (i.e. which
+class it belongs to).
+
+Without any knowledge about which virtual function can be called from which
+kernels compiler will have to make all virtual functions available to all
+kernels. That may not be desirable because some of those virtual functions could
+use optional kernel features and thus would propagate their use into kernels
+designed to be submitted to devices without support for those optional features.
```

Comment:
This is correct, but I think it would be clearer to explain it via the forbidden features.  For example:

> That may not be desirable because some of those virtual functions could use features that are prohibited in device code.

Reply:
Applied in ba37bc2db6614aaaac547eb58e8182fee78cc5ca